android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java

```diff
@@ -104,87 +106,57 @@ public void onError(BleError error) {
+            String json = multiCharacteristicsResponseJsonConverter.toJson(characteristicsResponse);
+            result.success(json);
+        } catch (BleError error) {
+            error.printStackTrace();
+            result.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error));
+        } catch (JSONException e) {
+            e.printStackTrace();
+            result.error(null, e.getMessage(), null);
+        }
     }
 
     private void getServices(String deviceId, final MethodChannel.Result result) {
-        adapter.getServicesForDevice(deviceId,
-                new OnSuccessCallback<Service[]>() {
-                    @Override
-                    public void onSuccess(final Service[] data) {
-                        try {
-                            result.success(new ServiceJsonConverter().toJson(data));
-                        } catch (JSONException e) {
-                            result.error(e.toString(), e.getMessage(), null);
-                        }
-                    }
-                }, new OnErrorCallback() {
-                    @Override
-                    public void onError(BleError error) {
-                        result.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error));
-                    }
-                });
+        try {
+            Service[] services = adapter.getServicesForDevice(deviceId);
+            result.success(serviceJsonConverter.toJson(services));
+        } catch (BleError error) {
+            error.printStackTrace();
+            result.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error));
+        } catch (JSONException e) {
+            e.printStackTrace();
+            result.error(null, e.getMessage(), null);
+        }
     }
 
     private void getCharacteristicsForService(Integer serviceId, final MethodChannel.Result result) {
-        adapter.getCharacteristicsForService(serviceId,
-                new OnSuccessCallback<Characteristic[]>() {
-                    @Override
-                    public void onSuccess(Characteristic[] data) {
-                        try {
-                            result.success(new CharacteristicJsonConverter().toJson(data));
-                        } catch (JSONException e) {
-                            result.error(e.toString(), e.getMessage(), null);
-                        }
-                    }
-                },
-                new OnErrorCallback() {
-                    @Override
-                    public void onError(BleError error) {
-                        result.error(String.valueOf(error.errorCode.code), error.reason, bleErrorJsonConverter.toJson(error));
-                    }
-                });
+        try {
+            Characteristic[] characteristics = adapter.getCharacteristicsForService(serviceId);
```

Comment:
Can characteristics be empty/null? If yes, what should happen? What happens if serviceId is unknown?

Reply:
Ad1. Characteristics can be empty, but cannot be null
Ad.2 If serviceId is null then BleError will be thrown.